### PR TITLE
ui:getChannelsUsage

### PR DESCRIPTION
**Pull request by gonzazoid**

### CODE REVIEW

**Comment by gonzazoid** [comment]
screenshot 2019-01-18 at 17 06 17

**Comment by sofabeat** [comment]
23 sec seems too long. Isn't it? Look like a possible bottleneck. @furkhat, please take a look.

File: data/offering.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package data
+
+func FilterOfferings(vs *[]*Offering, f func(*Offering) bool) []*Offering {
```
Comment (furkhat) [inline data/offering.go]: Don't send slices as a pointer.

File: data/session.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package data
+
+func FilterSessions(vs *[]*Session, f func(*Session) bool) []*Session {
```
Comment (furkhat) [inline data/session.go]: Don't send slices as a pointer. Functions repetitive with `FilterOfferings`.

File: doc/ui/rpc.md [hunk 3]
```diff
@@ -604,36 +604,36 @@ curl -X POST -H "Content-Type: application/json" --data '{"method": "ui_getClien
...
 // Request
-curl -X POST -H "Content-Type: application/json" --data '{"method": "ui_getChannelUsage", "params": ["qwert", "e66d8abd-c5e4-4ced-b9c3-fc3d61a911d0"], "id": 67}' http://localhost:8888/http
+curl -X POST -H "Content-Type: application/json" --data '{"method": "ui_getChannelsUsage", "params": ["qwert", ["e66d8abd-c5e4-4ced-b9c3-fc3d61a911d0"]], "id": 67}' http://localhost:8888/http
 
 // Result
 {
     "id": 67,
     "jsonrpc": "2.0",
-    "result": {
+    "result": [{
```
Comment (furkhat) [inline doc/ui/rpc.md]: I would add channel id somehow. Either as one more property in array or whole result as object with channel id as a key.

File: ui/channel.go [hunk 4]
```diff
@@ -198,32 +198,57 @@ func (h *Handler) GetAgentChannels(tkn string,
...
+	usage := make([]*Usage, len(channels))
+	for k, channel := range channels {
+		offering := data.FilterOfferings(&offerings, func(offer *data.Offering) bool { return channel.Offering == (*offer).ID })
+		if len(offering) != 1 {
+			return nil, ErrOfferingNotFound
+		}
+		channelSessions := data.FilterSessions(&sessions, func(session *data.Session) bool { return channel.ID == (*session).Channel })
+		usage[k] = newUsage(channel, offering[0], channelSessions)
+	}
+	return usage, nil
```
Comment (furkhat) [inline ui/channel.go]: Too noisy function, could be improved.

Comment (gonzazoid) [reply ui/channel.go]: I've just removed filtering. Now it's retrieve sorted data, which allows select sessions by getting slice on sorted sessions slice.

File: data/offering.go (hunk 1 above)
Comment (gonzazoid) [reply data/offering.go]: Removed

File: data/session.go (hunk 2 above)
Comment (gonzazoid) [reply data/session.go]: Removed

**Pull request merged**